Migrate and encapsulate logic from conversation skin directive to ratings-and-recommendations component.

ashutoshc8101 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #9749.
2. This PR does the following: Migrate and encapsulate logic from conversation skin directive to two new components: future-tutor-card and conversation-skin-final-summary components.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct


Cards animating:


https://user-images.githubusercontent.com/24855641/147766375-1f0850df-8cfb-4d3c-ae72-7f7f578248ff.mp4



Previous responses animating:

https://user-images.githubusercontent.com/24855641/141123971-e1bedfd1-9dfb-4102-987e-e656b262a8ed.mp4

Suggested exploration Ids working:

https://user-images.githubusercontent.com/24855641/141124065-ae22c788-ea31-4074-bebe-ca0ced71c446.mp4

Feedback working:

https://user-images.githubusercontent.com/24855641/141124120-a7acf5c9-4325-4167-a28f-9f4102bff273.mp4

Content shows up correctly on resizing window:

https://user-images.githubusercontent.com/24855641/141124197-2d483929-f602-4efd-b728-6be54452d6f8.mp4



## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ashutoshc8101 committed:
added final summary and future tutor card components

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
added unit tests

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
coverage fix

ashutoshc8101 commented:
@seanlip PTAL, In case you are interested.

EricZLou commented on core/templates/pages/exploration-player-page/exploration-player-page.module.ts:
```diff
@@ -106,7 +113,6 @@ class ExplorationPlayerPageModule {
 
 import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 import { downgradeModule } from '@angular/upgrade/static';
-import { ToastrModule } from 'ngx-toastr';
```
Not important, but we might as well move all of these imports to the top of the file, right?

ashutoshc8101 replied:
Importing platformBrowserDynamic and downgradeModule separately here is the pattern we are following for all *.module.ts files. The reason I can think of this is that code at end including these imports will be deleted once the page is competely migrated to angular.

EricZLou commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.ts:
```diff
@@ -1255,10 +1249,6 @@ angular.module('oppia').directive('conversationSkin', [
-
           // Returns whether the screen is wide enough to fit two
           // cards (e.g., the tutor and supplemental cards) side-by-side.
           $scope.canWindowShowTwoCards = function() {
```
Do we still need the function `canWindowShowTwoCards` to be here?

ashutoshc8101 replied:
Yes, it is used in `conversation-skin.directive.html` file. Line 15

EricZLou reviewed (commented):
Thanks @ashutoshc8101 -- just some small comments.

seanlip reviewed (commented):
Overall, seems like a nice approach -- thanks for looping me in @ashutoshc8101!

Please be sure to add videos for the embedded and non-embedded cases in the "Proof that changes are correct" section. We need to make sure things do not break.

EricZLou reviewed (commented):
LGTM but I want to take a look at the screen recording before approving. Thanks!

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 commented:
@EricZLou PTAL

ashutoshc8101 commented:
@seanlip @aishwary023 @gp201  PTAL

ashutoshc8101 committed:
nit

seanlip commented:
Hi @ashutoshc8101, I took a look at the video. Some additional suggestions for testing this:

- Try submitting multiple answers (maybe 5) and make sure the dropdown in the tutor card expands and collapses properly.
- Why is the green "Rating received" message coming up twice in the bottom right?
- Add an EndExploration interaction at the end, and add a couple of other exploration IDs to it. Make sure those show up correctly in the learner view.
- Try also resizing the window midway to see if everything still shows up correctly.
- It also seems to me that the animations are not working? When transitioning from one card to another, the tutor card should expand/contract smoothly so that it matches the height of the next card. To see this it might be good to make a really long content card followed by a really short one, and vice versa, so that you can check that the animation is happening correctly.

I think, with the above, you can have some confidence that things are working correclty.

Thanks!

EricZLou commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin-embed.directive.html:
```diff
@@ -34,6 +23,7 @@
           </supplemental-card>
         </div>
         <oppia-progress-nav (submit)="submitAnswerFromProgressNav()"
+                            [is-learn-again-button]="isLearnAgainButton"
```
Good catch!

ashutoshc8101 replied:
Thanks

DubeySandeep replied:
Maybe `learn-again-button-is-shown` similar to `submit-button-is-shown`? (or something around that i.e, not start with verb)

EricZLou reviewed (approved):
LGTM - thanks!

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

gp201 commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+    componentInstance.signIn();
+    tick();
+
+    expect(userService.getLoginUrlAsync).toHaveBeenCalled();
```
Also check the value of `windowRef.nativeWindow.location`

ashutoshc8101 replied:
I am using a mock for windowRef. So, changes won't reflect.

gp201 commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+      componentInstance.signIn();
+      tick();
+
+      expect(userService.getLoginUrlAsync).toHaveBeenCalled();
```
Also check the value of `windowRef.nativeWindow.location`

ashutoshc8101 replied:
I am using a mock for windowRef. So, changes won't reflect.

gp201 commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+
+    componentInstance.submitUserRating(5);
+
+    expect(learnerViewRatingService.submitUserRating).toHaveBeenCalled();
```
please use `tohaveBeenCalledWith()`

ashutoshc8101 replied:
done

ashutoshc8101 committed:
commit

ashutoshc8101 committed:
animations working

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
lint

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 commented:
@seanlip PTAL
With the new animations, off screen component is not required to calculate height of next card. Hence removed.

seanlip commented:
Will take a look in a bit, but that's great news -- the less code to maintain, the better. Thanks @ashutoshc8101! 

I'm a bit concerned that 4 CI checks are failing. Are these related to changes in your PR?

seanlip commented:
@ashutoshc8101 -- I looked at the first video and the animation when clicking the Back button is pretty choppy. Is it possible to make it smoother?

For the third video, the suggested explorations are supposed to appear in a horizontal row, not vertically.

Otherwise, it looks good. PTAL. Thanks!

ashutoshc8101 committed:
nits

ashutoshc8101 commented:
![Screenshot from 2021-11-12 16-39-20](https://user-images.githubusercontent.com/24855641/141466257-2b65091d-c6a2-4aef-8f75-0b3ed897492e.png)

@seanlip PTAL

Suggested exploration appears horizontal now.

I am having trouble fixing the choppy animation. Can we start the review process on this PR while I fix the animation in parallel?

ashutoshc8101 committed:
nit

seanlip commented on core/templates/components/smooth-height/smooth-height.component.ts:
```diff
@@ -0,0 +1,56 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Component for the smooth height change animation.
```
It's a bit hard to tell what this does from the name of the file. Perhaps smooth-height-animation.component.ts -- is there any standard in Angular for how to name "animation" components? Should we also be putting it in an animations folder? /cc @srijanreddy98

ashutoshc8101 replied:
Changed to `smooth-height-animation.component.ts`.
There is no concept of "animation" components in angular, this is a workaround that works when we want to animate to a dynamic height.

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin-embed.directive.html:
```diff
@@ -1,22 +1,11 @@
-    </div>
-  </div>
-</div>
+<!--
```
I thought you said we didn't need an off-screen preview any more? (ref: https://github.com/oppia/oppia/pull/14049#issuecomment-965307609)

ashutoshc8101 replied:
forgot to remove it from here.

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.html:
```diff
@@ -0,0 +1,155 @@
+    font-weight: 500;
+  }
+
+  .oppia-login-button.md-raised {
```
Should this and the next rule follow the .conversation-skin-... pattern, or will this mess up other components?

ashutoshc8101 replied:
These buttons have nothing specific to conversation skin directive except they are viewed inside it. It won't mess up other components.

seanlip replied:
But typically anything that's just prefixed with .oppia is defined in oppia.css and applies to the entire app. Until we change this codebase-wide, I suggest giving this some sort of prefix that clearly scopes it within this component (to other developers) if that is what the intent is.

(I'll also make a note to discuss in the next Angular team meeting what our approach to CSS naming is in light of the Angular migration, so that we have standard patterns to follow (a) during the hybrid phase, (b) post-migration.)

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+    userService = TestBed.inject(UserService);
+  });
+
+  it('should initialize', fakeAsync(() => {
```
Probably should be more like "should do X on initialize", e.g. "populate internal properties"

ashutoshc8101 replied:
Done

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+
+    componentInstance.togglePopover();
+
+    expect(componentInstance.feedbackPopOver.toggle).toHaveBeenCalled();
```
Here and below, is there any way to check the popover's final "state" rather than use toHaveBeenCalled()? The latter is weaker.

ashutoshc8101 replied:
No, because we are using a mock for ngbPopover.

seanlip replied:
OK. I suggest leaving a comment to explain this here (and elsewhere where toHaveBeenCalled() is used).

ashutoshc8101 replied:
Done

seanlip replied:
Ah -- sorry I was unclear. I think one more thing you'd want to explain (in these comments) is why we need to use a mock for the popover in the first place, since that seems to be a choice on our part.

seanlip replied:
Hi @ashutoshc8101 -- just flagging this comment since it hasn't been addressed yet. Thanks!

ashutoshc8101 replied:
Done

seanlip reviewed (changes requested):
Thanks @ashutoshc8101! Took a pass. 

And yup, let's add other codeowners.

seanlip commented:
@kevintab95 @DubeySandeep @srijanreddy98 @aishwary023 PTAL as codeowners. Thanks!

DubeySandeep commented on core/templates/components/smooth-height/smooth-height.component.ts:
```diff
@@ -0,0 +1,56 @@
+import { animate, style, transition, trigger } from '@angular/animations';
+
+@Component({
+  selector: 'oppia-smooth-height',
```
@ashutoshc8101 Any specific reason for creating this as a component and not directive (and use it as an attribute for a DOM/component)?

To me `oppia-smooth-height` doesn't sound like a view instead it sounds like a attribute which can affect a view.

ashutoshc8101 replied:
because content projection can't be done using a directive.

DubeySandeep replied:
But do we really need the internal html in this directive/componenet? Can't we find the height using the directive? (not sure!)

ashutoshc8101 replied:
Angular animations don't work with directives. They only work with components.

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -68,7 +69,6 @@
 /core/templates/services/UpgradedServices.ts @srijanreddy98 @ashutoshc8101
 /core/templates/hybrid-router-module-provider*.ts @srijanreddy98 @ashutoshc8101
 
-
```
We are expecting two newlines between project sections. Is this an expected change?

ashutoshc8101 replied:
Done by mistake, fixed.

DubeySandeep commented on core/templates/pages/exploration-player-page/exploration-player-page.module.ts:
```diff
@@ -68,6 +70,7 @@ import { MyHammerConfig, toastrConfig } from 'pages/oppia-root/app.module';
     LearnerLocalNavComponent,
     FeedbackPopupComponent,
     LearnerViewInfoComponent,
+    RatingsAndRecommendationsComponent,
```
Any specific reason for not having separate rating and recommendation components? (They looks pretty independent view to me)

ashutoshc8101 replied:
Not sure, but this component is not that big that it requires further sub-division.

DubeySandeep commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.ts:
```diff
@@ -0,0 +1,100 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Component for final summary to be displayed on conversation
```
Update the fileoverview*

ashutoshc8101 replied:
done

DubeySandeep commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.ts:
```diff
@@ -0,0 +1,100 @@
+  @Input() questionPlayerConfig;
+  @Input() inStoryMode: boolean;
+  @Input() storyViewerUrl!: string;
+  @Input() collectionSummary;
```
Do we need type here? (Ditto below)

ashutoshc8101 replied:
Added

DubeySandeep commented on core/templates/pages/exploration-player-page/learner-experience/tutor-card.component.html:
```diff
@@ -2,112 +2,113 @@
 <div class="oppia-learner-view-card"
      [ngClass]="{'animate-card-width': startCardChangeAnimation}"
```
Do we still need these animate class?

ashutoshc8101 replied:
Removed

DubeySandeep reviewed (changes requested):
@ashutoshc8101 Thanks for the PR, I've left a few comments, PTAL!

ashutoshc8101 committed:
suggestions

ashutoshc8101 committed:
name changed

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 commented:
@seanlip @DubeySandeep PTAL

ashutoshc8101 commented:
@gp201 PTAL

ashutoshc8101 committed:
unit test nits

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
lint

ashutoshc8101 committed:
lint

seanlip commented on .github/CODEOWNERS:
```diff
@@ -67,8 +68,6 @@
 /core/templates/services/request-interceptor.service*.ts @srijanreddy98 @ashutoshc8101
 /core/templates/services/UpgradedServices.ts @srijanreddy98 @ashutoshc8101
 /core/templates/hybrid-router-module-provider*.ts @srijanreddy98 @ashutoshc8101
```
I think you lost a (single) newline separating this from the next section?

ashutoshc8101 replied:
added

seanlip commented on core/templates/components/smooth-height/smooth-height-animation.component.ts:
```diff
@@ -0,0 +1,56 @@
+import { animate, style, transition, trigger } from '@angular/animations';
+
+@Component({
+  selector: 'oppia-smooth-height',
```
Is it worth calling this something like oppia-scroll-height-animation or oppia-scroll-height-animator, so that it's clear this is an "Animation" tag?

(Note that this would set a standard precedent for future animations.) 

What do other codebases do / what is standard practice?

(I ask because typically components are "things", and "smooth height" doesn't really sound like a "thing". That's actually the main underlying issue I see here.)

ashutoshc8101 replied:
`smooth-height` component is a workaround to achieve transition to a dynamic height. Typically, animations are not implemented using components. So, not sure what is the standard practice.

Changed selector to `oppia-scroll-height-animation`.

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.spec.ts:
```diff
@@ -677,17 +672,6 @@ describe('Conversation skin directive', function() {
   });
 
   describe('on initalization', function() {
-    it('should set user rating', function() {
```
For each deleted test here, can you point to its replacement in your PR (so that we're not reducing behavioural coverage) or explain why a replacement is not needed?

ashutoshc8101 replied:
moved to `ratings-and-recommendations.component.spec.ts` file.

ashutoshc8101 replied:
Done

seanlip reviewed (changes requested):
Thanks @ashutoshc8101! I took a full pass. Generally just small things left.

However, many CI checks are failing, and also it looks like the animation associated with the back button in the first video isn't fixed yet. You probably know this already but just thought I'd mention it.

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.spec.ts:
```diff
@@ -727,22 +711,6 @@ describe('Conversation skin directive', function() {
       expect($scope.collectionTitle).toBe(null);
     });
 
-    it('should call success alert message when ' +
-      'rating is updated successfully', function() {
```
moved to ratings-and-recommendations.component.spec.ts file.

ashutoshc8101 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.spec.ts:
```diff
@@ -1014,17 +982,6 @@ describe('Conversation skin directive', function() {
     });
   });
 
-  it('should open sign in page if user is not signed in' +
-    'when clicking on sign in button', fakeAsync(function() {
```
moved to ratings-and-recommendations.component.spec.ts file.

ashutoshc8101 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.spec.ts:
```diff
@@ -1382,16 +1339,6 @@ describe('Conversation skin directive', function() {
     expect(result).toBe('contentLabel2');
   });
 
-  it('should get feedback popover url when' +
-    'calling \'getFeedbackPopoverUrl\'', function() {
-    spyOn(urlInterpolationService, 'getDirectiveTemplateUrl')
-      .and.returnValue('/directiveTemplateUrl');
-
```
No required now, as we don't use feedback popover url to show popover.

ashutoshc8101 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.spec.ts:
```diff
@@ -1900,16 +1847,6 @@ describe('Conversation skin directive', function() {
     expect(displayCardSpy).toHaveBeenCalled();
   });
 
-  it('should submit user rating given rating value ' +
-    'when calling \'submitUserRating\'', function() {
```
moved to ratings-and-recommendations.component.spec.ts file.

ashutoshc8101 commented:
Reply to:
https://github.com/oppia/oppia/pull/14049#discussion_r752699499

Moved the css definations to oppia.css

https://github.com/oppia/oppia/pull/14049#discussion_r752700123

Added

ashutoshc8101 committed:
nits

ashutoshc8101 commented:
@seanlip PTAL

ashutoshc8101 committed:
add unit tests for smooth height component

ashutoshc8101 committed:
lint

DubeySandeep commented:
@ashutoshc8101 I've lefta a reply [here](https://github.com/oppia/oppia/pull/14049#discussion_r753760390), PTAL!

ashutoshc8101 commented:
@DubeySandeep PTAL

seanlip commented on core/templates/components/smooth-height/smooth-height-animation.component.ts:
```diff
@@ -0,0 +1,56 @@
+import { animate, style, transition, trigger } from '@angular/animations';
+
+@Component({
+  selector: 'oppia-smooth-height-animation',
```
Sorry -- could you go with `-animator` instead? On further thought "animation" doesn't really fit because it is not a thing, it refers to the technique of animating.

At least with `-animator` the idea is that the directive is doing or responsible for the animating.

ashutoshc8101 replied:
Done

seanlip commented on core/templates/css/oppia.css:
```diff
@@ -4842,3 +4844,21 @@ div.oppia-issues-learner-action {
+  margin-left: 10px;
+}
+
+.oppia-login-button {
```
Just to confirm: are these CSS rules used in more than one component?

If they are, keeping them here is fine; just leave review comments here pointing out the components that use them (there should be at least 2).

If not then they should be renamed and scoped into their corresponding components.

(I'm assuming it's the former, given previous discussion, but just wanted to see the proof so that we do this rigorously.)

ashutoshc8101 replied:
Moved .oppia-exp-summary-tiles-container to library-page.component.ts and other styles are used in more than 1 component.

seanlip reviewed (changes requested):
Only trivial comments, thanks @ashutoshc8101.

My main concern with this PR is the UI. Did you see my previous general top-level comment? I need to see the UI working properly before I can approve.

ashutoshc8101 committed:
fix e2e

DubeySandeep reviewed (approved):
LGTM for the changes in `.github/CODEOWNERS`

ashutoshc8101 committed:
suggestions

ashutoshc8101 commented:
Response to https://github.com/oppia/oppia/pull/14049#discussion_r753765467

oppia-feedback-popover is a separate component with its own spec file containing unit tests.
Not using a mock will make `ratings-and-recommendations` component dependent on `feedback-popover` which is against the principle of unit testing.

This seems obvious to me, do we still need to add this explanation as comment?

ashutoshc8101 commented:
@seanlip PTAL

ashutoshc8101 committed:
nit;

ashutoshc8101 committed:
lint

seanlip commented:
>> oppia-feedback-popover is a separate component with its own spec file containing unit tests.
Not using a mock will make ratings-and-recommendations component dependent on feedback-popover which is against the principle of unit testing.

> This seems obvious to me, do we still need to add this explanation as comment?

Yes, actually -- not all our frontend test files follow this principle yet, so probably worth calling it out. We don't always mock every sub-component.

There's also a bit of a philosophical debate here since we'd need to weigh this against the principle of preferring "equality checks" over "has this function been called". Ideally the mock state could be examined, and that would solve all the problems.

-----

Also, no other review comments besides the above and the question about the UI. Thanks!

gp201 reviewed (approved):
LGTM for codeowner files

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 committed:
animation fixed

ashutoshc8101 committed:
frotent tests fixed

ashutoshc8101 commented:
@seanlip PTAL

seanlip reviewed (changes requested):
Just one more comment. Otherwise everything LGTM!

ashutoshc8101 committed:
comment

ashutoshc8101 commented:
@seanlip PTAL

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -118,7 +118,9 @@ describe('Ratings and recommendations component', () => {
 
-    // Using a mock for popover don't allow us to check its final state.
+    // Using feedback popover directly here breaks the principle of
+    // indpendent unit tests. Hence, mock is used here.
```
"independent" is misspelled.

ashutoshc8101 replied:
fixed

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -118,7 +118,9 @@ describe('Ratings and recommendations component', () => {
-    // Using a mock for popover don't allow us to check its final state.
+    // Using feedback popover directly here breaks the principle of
+    // indpendent unit tests. Hence, mock is used here.
+    // The for popover don't allow us to check its final state.
```
What is "the for popover"?

ashutoshc8101 replied:
fixed

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -128,8 +130,10 @@ describe('Ratings and recommendations component', () => {
-    // Using a mock for popover don't allow us to check its final state.
-    // So, using a spy on function close instead.
+    // Using feedback popover directly here breaks the principle of
+    // indpendent unit tests. Hence, mock is used here.
```
Similar comments as above (both cases)

ashutoshc8101 replied:
fixed

ashutoshc8101 committed:
nits

ashutoshc8101 committed:
nit2

ashutoshc8101 commented:
@seanlip PTAL

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts:
```diff
@@ -120,7 +120,7 @@ describe('Ratings and recommendations component', () => {
     // Using feedback popover directly here breaks the principle of
     // independent unit tests. Hence, mock is used here.
-    // The mock  don't allow us to check its final state.
+    // The mock don't allow us to check its final state.
```
don't --> doesn't (here and elsewhere)

ashutoshc8101 replied:
done

ashutoshc8101 committed:
nit

ashutoshc8101 commented:
@srijanreddy98 PTAL

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 committed:
animation is smooth now

ashutoshc8101 committed:
nit

ashutoshc8101 committed:
coverage fix

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into conversation-skin-1

ashutoshc8101 committed:
nit

ashutoshc8101 reviewed (commented):
Took a final pass before merging

ashutoshc8101 committed:
e2e fix
